### ISSUE TITLE
ReverseProxyConventionBuilder is only return for one of the overload of MapReverseProxy

### ISSUE THREAD

**Issue by siby-george** [microsoft/reverse-proxy#860]
### What should we add or change to make your life better?
unseal ProxyRoute and Cluster or have a property bag to store deserialized object
### Why is this important to you?
we need to have some deserialized object like list of Roles which can be used in custom AuthZ requirement. currently we need to parse the roles from string every time

**Comment by Tratcher** [comment]
There are a few issues around extending the data model. Here's one of the older ones: https://github.com/microsoft/reverse-proxy/issues/87

Derivation is problematic for a few different reasons:
- The various configuration systems also have to be extended to populate the model
- You can only have one derivation, different components can't provide their own.

> we need to have some deserialized object like list of Roles which can be used in custom AuthZ requirement. currently we need to parse the roles from string every time

You re-parse them per request? You should at least be able to parse them once at startup and store the data. For example we've found ConditionalWeakTable to be a really useful way to store extra data associated with specific routes or clusters.

**Comment by siby-george** [comment]
> You re-parse them per request? You should at least be able to parse them once at startup and store the data. For example we've found ConditionalWeakTable to be a really useful way to store extra data associated with specific routes or clusters.

At least for now parsing is less expensive that lookup its just simple split.

What I was hope to do is derive customProxyRoute class from ProxyRoute and Custom Config provide will Populate addition prop and return CustomProxyRoute instead of ProxyRoute and do pattern match to see if this CustomProxyRoute and get extended properties if matched. Do you see any problem with this approach if we override Equals.
I will explorer more on ConditionalWeakTable but if I am not wrong it will add one addition lookup.
call back during route build will a flexible extension point.

**Comment by Tratcher** [comment]
Maintaining a custom config provider will be pretty expensive, that's one of the more complex parts of the code base. Having to downcast to consume the custom types isn't a great pattern either. The other place I'd be concerned is anywhere that clones the route objects like these extensions:
https://github.com/microsoft/reverse-proxy/blob/7d26dffc565ccaca7c44a46002e819df1a8b18b2/src/ReverseProxy/Abstractions/Config/ProxyRouteTransformExtensions.cs#L40

I'm still primarily concerned about how this approach doesn't scale. E.g. if you wanted to do this with the service fabric or kubernetes configuration providers that would be a lot more code to fork and maintain. Kubernetes even json serializes the config across processes, that's not a flow that works with derived types. This approach also couldn't be used with libraries such as if someone provided a custom load balancer implementation that needed its own config properties on Cluster or Destination. If they provide a derived cluster than nobody else can, there's no way to combine derivations.

In short, derivation might work in some narrow cases, but it's not a good general purpose solution to extending the config model. We need to find an approach that does generalize better. The metadata collection model we have today avoids most of these issues, but has some of its own usability concerns that you've raised.

**Comment by siby-george** [comment]
Appreciate you quick responses :) 
I agree it won't work if object is clone. I also not realized we have ReverseProxyConventionBuilder.ConfigureEndpoints to add parsed object to Endpoint metadata directly.  ReverseProxyConventionBuilder is only return for one of the overload
https://github.com/microsoft/reverse-proxy/blob/7d26dffc565ccaca7c44a46002e819df1a8b18b2/src/ReverseProxy/Service/DynamicEndpoint/ReverseProxyConventionBuilder.cs#L58
Looking in that direction hope this will resolve my problem

**Comment by Tratcher** [comment]
> ReverseProxyConventionBuilder is only return for one of the overload

Ah, that we should fix.
https://github.com/microsoft/reverse-proxy/blob/7d26dffc565ccaca7c44a46002e819df1a8b18b2/src/ReverseProxy/Configuration/ReverseProxyIEndpointRouteBuilderExtensions.cs#L23-L37
